### ISSUE TITLE
https://pjreddie.com/darknet/hardware-guide/ Faulty Hardware Guide ? 

### ISSUE THREAD

**Issue by dexception** [pjreddie/darknet#754]
CPU: Intel Core i7-5820K is not in any way capable of handling 4xTitan X GPU.
Minimum 30% bottleneck and wastage ! This architecture is only capable of handling 2xTitan X GPU.

Please share your thoughts and recommend an architecture for 4xTitan X GPU.

Thanks.

**Comment by AlexeyAB** [comment]
Are you talking about low number of PCIe-lanes, only 28 instead of 40 as in High-end desktop CPU? https://ark.intel.com/products/82932/Intel-Core-i7-5820K-Processor-15M-Cache-up-to-3_60-GHz

Or do you see only 70% of GPU-usage, because GPUs wait for data augmentation processing on CPU?
What GPU-usage can you see by using this fork with `OPENCV=1` in the Makefile? https://github.com/AlexeyAB/darknet

There is optimized data augmentation 3.5x times using OpenCV functions, so compile it with OPENCV=1 (and if the server doesn't have monitor - run training with `-dont_show` flag).
Also it supports Tensor Cores - set `CUDNN_HALF=1` in the Makefile. I trained model on 1 x Tesla V100 with Tensor Cores ~100 000 Gflops on CPU with 8 CPU-Cores, and **GPU-usage was about ~95%**: https://github.com/pjreddie/darknet/issues/704

* 4 x Titan X (GM200) is about 4 x 7000 = 28 000 Gflops.
* 1 x Tesla V100 with Tensor Cores ~100 000 Gflops

![38811862-6538f0f8-4193-11e8-80df-6eb96f29059b](https://user-images.githubusercontent.com/4096485/39656632-00e25920-500a-11e8-9f7a-8325b48f4459.png)

**Comment by dexception** [comment]
Yes i am talking about PCIe-lanes. You need 40 lanes to fully utilize 4xTitan X GPU SLI. 
Something like a Dual Xeon would work way better.

**Comment by AlexeyAB** [comment]
@dexception There is enough of Intel Core i9-7900X X-series 10(20)-Cores - **1000$ with 44 PCIe-lanes** - it is much cheaper than 2 x Xeon: https://ark.intel.com/products/123613/Intel-Core-i9-7900X-X-series-Processor-13_75M-Cache-up-to-4_30-GHz

But in practice, 1 x Tesla V100 with Tensor Cores (100 000 Gflops) processes more images per second so **transmits a larger data stream over PCIe** than 4 x Titan X GM200 (4 x 7000 Gflops). multi-GPUs are synchronized for each 100 iteratons only.
And there is enough of 16 lanes of PCIe 3.0 for ~95% GPU-usage of Tesla V100.

**Comment by uday60** [comment]
After having a discussion with the hardware engineer who has made systems for Cisco he confirmed there will be a bottleneck for 4x 1080ti GPU SLI architecture. And you will need more PCIe express lanes. 
1 CPU is just not enough. http://thebottlenecker.com/

Other issue...
I am training my model in 64 batch and 32 subdivisions. Also after using nivida-smi -a it shows i am using the cores but not all the memory. Can i further optimize this ?
    Utilization
        Gpu                         : 96 %
        Memory                      : 35 %
        Encoder                     : 0 %
        Decoder                     : 0 %

**Comment by dexception** [comment]
@AlexeyAB 
Tesla V100  PCI Express interfaces PCI Express 3.0 ×16, Lane and polarity reversal supported
https://images.nvidia.com/content/tesla/pdf/Tesla-V100-PCIe-Product-Brief.pdf

Suppose your using 4 GPUs together with 1 CPU Why won't there be not bottleneck when CPU only supports 16 lanes ? 
4 GPU will need 4x16PCIe express lanes = 80 lanes.. Intel Xeon chip has 40 lanes and Dual Xeon on board will be 80 lanes.  

Entry level Xeon is cheaper than  Intel Core i9-7900X.

**Comment by AlexeyAB** [comment]
@dexception Darknet doesn't use explicit duplex data transfer, so 2 x DMA (polarity reversal) isn't used simultaneously in the Darknet. 
How many percent have you been able to increase performance using 40 PCIe-lanes instead of 16 - for 4 x Titan X?

Yes, there is Xeon Silver 4112 - $483 with 48 lanes that is cheaper than Core i9-7900X: https://ark.intel.com/products/123551/Intel-Xeon-Silver-4112-Processor-8_25M-Cache-2_60-GHz

**Comment by dexception** [comment]
@AlexeyAB You should have a bottleneck of 20%-30%.

The fastest CPU will still be weak. You will need dual processor to manage 80 lanes for 4 way SLI.
http://thebottlenecker.com

**Comment by AlexeyAB** [comment]
@dexception 

* What GPU-usage and PCIe-usage do you get on the 4 x Titan X and what CPU do you use? May be you are right. Please, let's figure it out.

> You should have a bottleneck of 20%-30%.(Only in case of SLI)

* Only if your soft uses 100% of PCIe bandwidth. Otherwise the slowdown will be 30% of the X % that the data transfer CPU<->GPU takes in the Darknet for 40 lanes, or slowdown 75% of the X % for only 16 lanes - In comparison with the system having 64 lanes.
    Weights are uploaded to the GPU once. We should transfer from CPU to GPU only images: `416*416*3*64*sizeof(float) = 126 MB` for 1 iteration (1 batch for 64 images). So on PCIe 3.0 16x (16GB/sec specification, 14 GB/sec actual), data transfer CPU->GPU takes `126MB / 14000GB/s =` **~0.009 sec**. But 4 x Titan X processes one batch more than 1 sec. So data transfer takes `X = 100% * 0.009 / 1 = 0.9%` or less. So for 16 PCIe-lanes slowdown in the bottleneck will be about `75% * 0.9% ~= 0.675%`, **therefore you can win about 0.00675 sec from 1 sec by using 64 PCIe-lanes instead of 16 lanes.**

    Data transfer for calculating delta during training - CPU->GPU and GPU->CPU is: `sizeof(float) * (52*52*255 + 26*26*255 + 13*13*255) x 2(GPU->CPU->GPU) = 4 * 3549 * 255 * 2 = 6.9 MB` is about `~0.0005 sec` on PCIe 3.0 16-lanes.

    **You can save up to not more than `0.0095 seconds` for 1 batch (64 images) that usually takes more than `1 second` in the Darknet-Yolo-v3 (416x416) by using any soft/hardware optimizations of data transfer.**

* **SLI can't be used on CUDA.** SLI (that is used for graphics) isn't the same as [GPU-Direct2.0 P2P](https://developer.nvidia.com/gpudirect) (that is used for CUDA and is only supported on Quadro/Tesla). `CUDA Peer-to-Peer transactions do not use the SLI mechanism/path/cable.`: https://devtalk.nvidia.com/default/topic/845381/cuda-programming-and-performance/peer-to-peer-between-different-type-of-cards/post/4572106/#4572106

* Darknet doesn't use CUDA P2P-data transfer between GPUs. Darknet synchronizes weights from GPUs on CPU for each 100 iterations only.

**Comment by dexception** [comment]
@AlexeyAB 
I got my answer. The hardware guy was correct there is a bottleneck but it is negated at the application level. What would be the training time difference for a single GTX 1080 ti vs GTX 4x1080ti ? 

Thanks

**Comment by dexception** [comment]
@AlexeyAB Any replies ?

**Comment by powelllens** [comment]
@dexception 
I have two setups for training, one GTX1060 and another with 2x GTX1080ti. Comparing training only, the two GTX1080ti are more ore less ~6times faster than my development setup with the GTX1060.

random=1
yolov3 cfg
net 608x608
~1000 images
(2x GTX 1080ti --> 1h for 1000 iterations)

Single or double GTX1080ti training is close to double the speed. Only rough estimations coparing the times the weightfiles are created.
So in your case 4x1080ti compared to one, would be 4x times the training speed. For private use I have no idea why someone should spend the money for 4x GTX1080ti in times of blockchain mining, this is way to expensive, and way to energie hungry. Also even if the Titan V is ~3 times faster with CUDNN_HALF, compared to the price, the speed/$ is the same or even worse copared to 3x GTX1080ti.

I hope this helped.
